File: core-ops-service/app/pairing.py
"""Swiss pairing engine using NetworkX maximum weight matching (Blossom algorithm).

The engine constructs a weighted graph where:
  - Each active player is a node
  - Edge weights encode Swiss pairing preferences
  - Edges are REMOVED between previous opponents (hard constraint)
  - A dummy "BYE" node is added for odd player counts

Edge Weight Scheme (exponential):
  Uses W = base^(max_diff - actual_diff) where base = max(N, 10).
  This guarantees that same-score-group pairings always dominate
  cross-group pairings in the global matching — the combined weight
  of any number of cross-group pairings cannot exceed a single
  same-group pairing when base >= N.

  Previous opponents:   Edge REMOVED entirely
  Bye Dummy Node:       weight = rank_order (lower ranked → higher weight)
                        Always less than any player-vs-player edge.
                        Edge removed if player already had a bye

Determinism:
  Edge list sorted by (min(p1, p2), max(p1, p2)) before matching.
"""

from __future__ import annotations

import networkx as nx

from .models import PairingOutput, PairingRequest, PlayerInput

# Sentinel ID for the bye dummy node
BYE_SENTINEL = "__BYE__"


def _build_opponent_set(request: PairingRequest) -> dict[str, set[str]]:
    """Build a lookup of previous opponents for each player."""
    opponents: dict[str, set[str]] = {}
    for p in request.players:
        opponents.setdefault(p.id, set())
        for opp in p.opponents:
            opponents[p.id].add(opp)
    # Also use explicit pairing history
    for rec in request.previous_pairings:
        opponents.setdefault(rec.p1, set()).add(rec.p2)
        opponents.setdefault(rec.p2, set()).add(rec.p1)
    return opponents


def _compute_edge_weight(
    p1: PlayerInput, p2: PlayerInput, n_players: int
) -> int:
    """Compute the edge weight between two real players.

    Uses an exponential scheme: W = base^(max_diff - actual_diff)
    where base = max(n_players, 10).

    This guarantees same-score-group pairings always dominate cross-group
    pairings in the global matching. When base >= N, any single same-group
    edge outweighs the sum of all possible cross-group edges, so the
    Blossom algorithm will never sacrifice a same-group pairing to improve
    lower groups.

    Higher weight = more desirable pairing.
    """
    point_diff = abs(p1.match_points - p2.match_points)
    base = max(n_players, 10)
    # max_diff upper-bound: in a tournament, the maximum match point spread
    # is bounded by 3 × (n_players - 1), but we cap it for practicality.
    # A value of 18 (6 rounds × 3 pts) covers any realistic tournament.
    max_diff = 18
    return base ** (max_diff - point_diff)


def generate_pairings(request: PairingRequest) -> tuple[list[PairingOutput], str | None]:
    """Generate Swiss pairings using Blossom algorithm.

    Args:
        request: PairingRequest with players, previous pairings, and config.

    Returns:
        Tuple of (list of PairingOutput, bye_player_id or None).

    Raises:
        ValueError: If pairing is impossible (e.g., all players have faced each other).
    """
    # Filter to active (non-dropped) players
    active_players = [p for p in request.players if not p.is_dropped]

    if len(active_players) < 2:
        raise ValueError("Need at least 2 active players to generate pairings.")

    opponent_sets = _build_opponent_set(request)
    needs_bye = len(active_players) % 2 == 1

    # Build graph
    G = nx.Graph()

    # Add player nodes
    for p in active_players:
        G.add_node(p.id)

    # Add bye node if needed
    if needs_bye and request.config.allow_byes:
        G.add_node(BYE_SENTINEL)

    # Sort players by match_points descending for rank ordering
    sorted_players = sorted(active_players, key=lambda p: p.match_points, reverse=True)
    player_rank = {p.id: rank for rank, p in enumerate(sorted_players)}
    player_map = {p.id: p for p in active_players}

    # Add edges between all player pairs (excluding previous opponents)
    player_ids = [p.id for p in active_players]
    for i in range(len(player_ids)):
        for j in range(i + 1, len(player_ids)):
            pid1 = player_ids[i]
            pid2 = player_ids[j]

            # Hard constraint: no rematches
            if pid2 in opponent_sets.get(pid1, set()):
                continue

            weight = _compute_edge_weight(
                player_map[pid1], player_map[pid2], len(active_players)
            )
            # Ensure deterministic ordering of the edge tuple
            edge = tuple(sorted([pid1, pid2]))
            G.add_edge(edge[0], edge[1], weight=weight)

    # Add bye edges
    if needs_bye and request.config.allow_byes:
        for p in active_players:
            if p.has_bye:
                # Player already had a bye — don't add edge
                continue
            # Lower-ranked players get higher weight for bye (more likely to get bye)
            bye_weight = player_rank[p.id] + 1  # rank 0 → weight 1, rank N → weight N+1
            edge = tuple(sorted([p.id, BYE_SENTINEL]))
            G.add_edge(edge[0], edge[1], weight=bye_weight)

    # Run Blossom algorithm
    matching = nx.max_weight_matching(G, maxcardinality=True)

    if not matching:
        raise ValueError("Failed to generate pairings — no valid matching found.")

    # Extract results
    pairings: list[tuple[str, str]] = []
    bye_player: str | None = None

    for u, v in matching:
        # Normalize order
        a, b = (u, v) if u < v else (v, u)

        if BYE_SENTINEL in (a, b):
            bye_player = a if b == BYE_SENTINEL else b
        else:
            pairings.append((a, b))

    # Sort pairings by combined match points descending (highest scoring tables first)
    pairings.sort(
        key=lambda pair: player_map[pair[0]].match_points + player_map[pair[1]].match_points,
        reverse=True,
    )

    # Assign table numbers (1-indexed, highest scoring pair = table 1)
    pairing_outputs = [
        PairingOutput(
            p1=p1,
            p2=p2,
            table_number=table_num,
        )
        for table_num, (p1, p2) in enumerate(pairings, start=1)
    ]

    return pairing_outputs, bye_player

File: lib/core-ops/pairing.ts
/**
 * Core Ops — Swiss Pairing Engine using Edmonds' Blossom algorithm.
 *
 * Port from Python pairing.py. Uses `edmonds-blossom` npm package
 * instead of NetworkX for maximum weight matching.
 *
 * Edge Weight Scheme:
 *   Same score group:     BASE = 1000
 *   Cross score group:    BASE = 1000 - |point_diff|² × 10  (quadratic penalty)
 *   Previous opponents:   Edge REMOVED entirely
 *   Bye Dummy Node:       weight = rank_order × 100 + 1 (lower ranked → higher weight)
 *                          Edge removed if player already had a bye
 */

import blossom from "edmonds-blossom";

// --- Types ---

export interface PlayerInput {
  id: string;
  matchPoints: number;
  isDropped: boolean;
  hasBye: boolean;
  /** IDs of previously faced opponents */
  opponents: string[];
}

export interface PairingRecord {
  p1: string;
  p2: string;
  round: number;
}

export interface PairingConfig {
  allowByes: boolean;
}

export interface PairingOutput {
  p1: string;
  p2: string;
  tableNumber: number;
}

export interface PairingResult {
  pairings: PairingOutput[];
  byePlayer: string | null;
}

// --- Constants ---

const BYE_SENTINEL = "__BYE__";

// --- Internal Helpers ---

function buildOpponentSets(
  players: PlayerInput[],
  previousPairings: PairingRecord[]
): Map<string, Set<string>> {
  const opponents = new Map<string, Set<string>>();

  for (const p of players) {
    if (!opponents.has(p.id)) opponents.set(p.id, new Set());
    for (const opp of p.opponents) {
      opponents.get(p.id)!.add(opp);
    }
  }

  for (const rec of previousPairings) {
    if (!opponents.has(rec.p1)) opponents.set(rec.p1, new Set());
    if (!opponents.has(rec.p2)) opponents.set(rec.p2, new Set());
    opponents.get(rec.p1)!.add(rec.p2);
    opponents.get(rec.p2)!.add(rec.p1);
  }

  return opponents;
}

function computeEdgeWeight(p1: PlayerInput, p2: PlayerInput): number {
  const pointDiff = Math.abs(p1.matchPoints - p2.matchPoints);
  if (pointDiff === 0) return 1000;
  // Quadratic penalty: big gaps cost disproportionately more than small ones.
  // This prevents Blossom from choosing one large gap (e.g. 4-0 vs 1-3)
  // when multiple small floats (e.g. 3× one-bracket floats) yield the same total.
  return Math.max(1, 1000 - pointDiff * pointDiff * 10);
}

// --- Main Function ---

/**
 * Generate Swiss pairings using Blossom algorithm.
 *
 * @param players Active players to pair
 * @param previousPairings All previous pairings in this tournament
 * @param config Configuration (allow byes etc.)
 * @returns PairingResult with pairings and bye player
 * @throws Error if pairing is impossible
 */
export function generatePairings(
  players: PlayerInput[],
  previousPairings: PairingRecord[] = [],
  config: PairingConfig = { allowByes: true }
): PairingResult {
  // Filter to active (non-dropped) players
  const activePlayers = players.filter((p) => !p.isDropped);

  if (activePlayers.length < 2) {
    throw new Error("Need at least 2 active players to generate pairings.");
  }

  const opponentSets = buildOpponentSets(activePlayers, previousPairings);
  const needsBye = activePlayers.length % 2 === 1;

  // Build node list: all active players + optional bye sentinel
  const nodeIds: string[] = activePlayers.map((p) => p.id);
  if (needsBye && config.allowByes) {
    nodeIds.push(BYE_SENTINEL);
  }

  // Create index mapping: playerID → integer node index
  const idToIndex = new Map<string, number>();
  for (let i = 0; i < nodeIds.length; i++) {
    idToIndex.set(nodeIds[i], i);
  }

  const playerMap = new Map(activePlayers.map((p) => [p.id, p]));

  // Sort players by match points descending for rank ordering
  const sortedPlayers = [...activePlayers].sort(
    (a, b) => b.matchPoints - a.matchPoints
  );
  const playerRank = new Map(sortedPlayers.map((p, rank) => [p.id, rank]));

  // Build edge list for Blossom: [nodeIdx1, nodeIdx2, weight]
  const edges: number[][] = [];

  // Add edges between all player pairs (excluding previous opponents)
  for (let i = 0; i < activePlayers.length; i++) {
    for (let j = i + 1; j < activePlayers.length; j++) {
      const pid1 = activePlayers[i].id;
      const pid2 = activePlayers[j].id;

      // Hard constraint: no rematches
      if (opponentSets.get(pid1)?.has(pid2)) continue;

      const weight = computeEdgeWeight(
        playerMap.get(pid1)!,
        playerMap.get(pid2)!
      );
      edges.push([idToIndex.get(pid1)!, idToIndex.get(pid2)!, weight]);
    }
  }

  // Add bye edges
  if (needsBye && config.allowByes) {
    const byeIdx = idToIndex.get(BYE_SENTINEL)!;
    // Scale bye weight so the lowest-ranked player's preference always
    // dominates over any possible improvement in player-player pairing.
    // Gap of N × 1000 per rank ensures bye assignment is unbreakable:
    // max pairing improvement ≤ (N/2) × 999 < N × 1000 = one rank gap.
    const BYE_WEIGHT_BASE = activePlayers.length * 1000;
    for (const p of activePlayers) {
      if (p.hasBye) continue; // Player already had a bye
      // Lower-ranked players get higher weight for bye
      const byeWeight = (playerRank.get(p.id) ?? 0) * BYE_WEIGHT_BASE + 1;
      edges.push([idToIndex.get(p.id)!, byeIdx, byeWeight]);
    }
  }

  // Run Blossom algorithm
  const mate: number[] = blossom(edges, true);

  if (!mate || mate.every((m) => m === -1)) {
    throw new Error("Failed to generate pairings — no valid matching found.");
  }

  // Extract results from mate array
  const pairingPairs: [string, string][] = [];
  let byePlayer: string | null = null;
  const seen = new Set<number>();

  for (let i = 0; i < mate.length; i++) {
    if (mate[i] === -1 || seen.has(i)) continue;
    const j = mate[i];
    seen.add(i);
    seen.add(j);

    const idA = nodeIds[i];
    const idB = nodeIds[j];

    if (idA === BYE_SENTINEL) {
      byePlayer = idB;
    } else if (idB === BYE_SENTINEL) {
      byePlayer = idA;
    } else {
      // Normalize order (lower ID first for determinism)
      const [lo, hi] = idA < idB ? [idA, idB] : [idB, idA];
      pairingPairs.push([lo, hi]);
    }
  }

  // --- Completeness validation ---
  // Verify every active player is covered by either a pairing or the bye.
  // This catches partial matches (blossom returns mate[i] === -1 for some
  // players) AND the case where mate is shorter than nodeIds (completely
  // isolated players whose high indices fall outside mate's range).
  const coveredIds = new Set<string>();
  for (const [a, b] of pairingPairs) {
    coveredIds.add(a);
    coveredIds.add(b);
  }
  if (byePlayer) coveredIds.add(byePlayer);

  const uncoveredPlayers = activePlayers.filter((p) => !coveredIds.has(p.id));

  // When byes are disabled and the player count is odd, exactly 1 player
  // will inevitably be left out — that's an expected limitation, not a bug.
  const expectedUncovered = needsBye && !config.allowByes ? 1 : 0;

  if (uncoveredPlayers.length > expectedUncovered) {
    const uncoveredIds = uncoveredPlayers.map((p) => p.id);
    throw new Error(
      `Failed to generate a complete pairing: ${uncoveredIds.length} player(s) could not be paired ` +
        `without a rematch: [${uncoveredIds.join(", ")}]. This usually happens when the field has ` +
        `asymmetric match histories (drops/re-adds, late entries).`
    );
  }

  // Sort pairings by combined match points descending (table 1 = highest scoring)
  pairingPairs.sort((a, b) => {
    const scoreA =
      (playerMap.get(a[0])?.matchPoints ?? 0) +
      (playerMap.get(a[1])?.matchPoints ?? 0);
    const scoreB =
      (playerMap.get(b[0])?.matchPoints ?? 0) +
      (playerMap.get(b[1])?.matchPoints ?? 0);
    return scoreB - scoreA;
  });

  // Assign table numbers (1-indexed)
  const pairings: PairingOutput[] = pairingPairs.map(([p1, p2], idx) => ({
    p1,
    p2,
    tableNumber: idx + 1,
  }));

  return { pairings, byePlayer };
}
